[LTO] Introduce a helper function collectImportStatistics (NFC)

kazutakahirata opened the pull request:
This patch introduces a helper function collectImportStatistics.  The
new function computes statistics of imports for
ComputeCrossModuleImport and dumpImportListForModule with no
functional change.

The background is as follows.  I'm planning to reduce the memory
footprint of ThinLTO indexing by changing ImportMapTy, the data
structure used for an import list.  The new list will be a hash set of
tuples (SourceModule, GUID, ImportType) represented in a space
efficient manner.  That means that obtaining statistics like the
number of definitions per source module requires us to go through the
entire import list (for a given destination module).

Introducing a helper function now makes the callers more independent
of the underlying data structures used in ImportMapT.

kazutakahirata committed:
[LTO] Introduce a helper function summarizeImports (NFC)

This patch introduces a helper function summarizeImports.  The new
function computes a summary of imports for ComputeCrossModuleImport
and dumpImportListForModule with no functional change.

The background is as follows.  I'm planning to reduce the memory
footprint of ThinLTO indexing by changing ImportMapTy, the data
structure used for an import list.  The new list will be a hash set of
tuples (SourceModule, GUID, ImportType) represented in space efficient
manner.  That means that obtaining a summary like the number of
definitions per source module requires us to go through the entire
import list (for a given destination module).

Introducing a helper function now makes the callers more independent
of the underlying data structures used in ImportMapT.

jvoung commented on llvm/lib/Transforms/IPO/FunctionImport.cpp:
```diff
@@ -1082,21 +1082,29 @@ numGlobalVarSummaries(const ModuleSummaryIndex &Index,
+
+  for (const auto &[FromModule, GUIDs] : ImportList.getImportMap()) {
+    for (const auto &[GUID, Type] : GUIDs) {
+      ImportSummary &Entry = Histogram[FromModule];
```
Can the Entry lookup be lifted to the outer loop (only depending on FromModule)?

kazutakahirata replied:
Done.  Thanks!

jvoung commented on llvm/lib/Transforms/IPO/FunctionImport.cpp:
```diff
@@ -1082,21 +1082,29 @@ numGlobalVarSummaries(const ModuleSummaryIndex &Index,
-numGlobalVarSummaries(const ModuleSummaryIndex &Index,
-                      const FunctionImporter::FunctionsToImportTy &ImportMap,
-                      unsigned &DefinedFS) {
+struct ImportSummary {
```
nit: I don't feel too strongly, but Summary is also used in the sense of GlobalVarSummary, FunctionSummary, etc. and so I wonder if using a different term like "Statistics" might be better.

teresajohnson replied:
Agree, the word summary ends up being overloaded - I like "Statistics" (here and in the uses where the stats get dumped later on) better.

kazutakahirata replied:
I've renamed `ImportSummary` to `ImportStatistics`.  Also, I've renamed `summarizeImports` to `collectImportStatistics`.  Thanks!

kazutakahirata committed:
Rename ImportSummary to ImportStatistics.

Rename summarizeImports to collectImportStatistics.

Move a map lookup outside the inner loop.

kazutakahirata commented:
I've updated the title, commit message, and patch proper.  PTAL.  Thanks!

jvoung commented on llvm/lib/Transforms/IPO/FunctionImport.cpp:
```diff
@@ -1217,21 +1225,19 @@ void llvm::ComputeCrossModuleImport(
-          numGlobalVarSummaries(Index, Src.second, DefinedFS);
-      LLVM_DEBUG(dbgs() << " - " << DefinedFS << " function definitions and "
-                        << Src.second.size() - NumGVSPerMod - DefinedFS
+    for (const auto &[SrcModName, Summary] : Histogram) {
```
small nit -- update the local variable names from "Summary" too?

kazutakahirata replied:
Done in both callers.  Thanks!

kazutakahirata committed:
Rename Summary to Stats.

teresajohnson commented on llvm/lib/Transforms/IPO/FunctionImport.cpp:
```diff
@@ -1082,21 +1082,29 @@ numGlobalVarSummaries(const ModuleSummaryIndex &Index,
+  unsigned Count = 0;
+};
+
+// Compute import summary for each source module in ImportList.
```
nit: s/summary/statistics/

kazutakahirata replied:
Done.  Thanks!

teresajohnson reviewed (approved):
lgtm with a comment fix

kazutakahirata committed:
Update a comment.
